### ISSUE TITLE
Operation vs Variable

### ISSUE THREAD

**Issue by YingboMa** [SciML/ModelingToolkit.jl#433]
I think it might be a good idea to use `Operation`s everywhere. This should solve #415. The problem is that the states are variables but the equations can have an `Operation` that behaves like a "variable", i.e.
```julia
julia> sys = ODESystem(eqs, t, [u,Dk(u)], [k])
ODESystem(Equation[Equation(derivative(u(t), t), -k * u(t)), Equation(derivative(derivative(u(t), k), t), -(u(t)) - k * derivative(u(t), k))], t, Variable[u, uˍk], Variable[k], Base.RefValue{Array{Expression,1}}(Expression[]), Base.RefValue{Array{Expression,2}}(Array{Expression}(undef,0,0)), Base.RefValue{Array{Expression,2}}(Array{Expression}(undef,0,0)), Base.RefValue{Array{Expression,2}}(Array{Expression}(undef,0,0)), Symbol("##ODESystem#294"), ODESystem[])

julia> sys.eqs
2-element Array{Equation,1}:
 Equation(derivative(u(t), t), -k * u(t))
 Equation(derivative(derivative(u(t), k), t), -(u(t)) - k * derivative(u(t), k))

julia> sys.states
2-element Array{Variable,1}:
 u
 uˍk
```

Though this could break A LOT OF internal assumptions, and it will take some time to do this refactoring.

**Comment by YingboMa** [comment]
Using `Operation`s everywhere could make the order lowering code a lot nicer, too.

**Comment by ChrisRackauckas** [comment]
Operations everywhere might make a lot of things nicer. I know @isaacsas was thinking about it. I think someone just has to do it for us to see whether it's better. I think it would probably be a change for the good.

**Comment by isaacsas** [comment]
This would be great, assuming it would not cause performance issues when checking equality or such. Something like https://github.com/SciML/ModelingToolkit.jl/pull/312 would be good to have too for Dictionary usage, which gives a big speed up for large systems over having to search array mappings. (Though testing it out right now I guess Operation to integer mappings in Dicts already work, which might be all I would need.)

**Comment by ChrisRackauckas** [comment]
In a similar vein, I wonder if we even need `Constant(x)`, since it can always be represented as the Operation `identity(x)`

**Comment by ChrisRackauckas** [comment]
I don't think we can fix https://github.com/SciML/ModelingToolkit.jl/pull/312 all of the way without fixing this.

**Comment by shashi** [comment]
Closed by #518
